docs/src/sparse_int_set.md

```diff
@@ -0,0 +1,9 @@
+# DataStructures.SparseIntSet
+
+Implementation of a __Sparse Integer Set__, for background see [Sparse Sets](https://www.computist.xyz/2018/06/sparse-sets.html).
+Only positive non-zero `Int`s are allowed inside the set. 
+The idea is to have one **packed** `Vector` storing all the `Int`s contained in the set as to allow for fast iteration, and a sparse, paged **reverse** `Vector` with the position of a particular `Int` inside the **packed** `Vector`. This allows for very fast iteration, insertion and deletion of indices.
+Most behavior is similar to a normal `IntSet`, however `collect`, `first` and `last` are with respected to the **packed** vector, in which the ordering is not guaranteed. 
+The **reverse** `Vector` is paged, meaning that it is a `Vector{Vector{Int}}` where each of the `Vector{Int}`s has the length of one memory page of `Int`s. Every time an index that was not yet in the range of the already present pages, a new one will be created and added to the **reverse**, allowing for dynamical growth. 
+If all the indices on a particular page are deleted from the set, it will not automatically get cleaned up for performance reasons. The `cleanup!` method is provided to facilitate the reclaiming of memory of a page when all it's corresponding indices have been deleted from the set.
```

Comment:
Having to manually trigger cleanup bothers me.
Can we maybe trigger this to run after every N remove operations or something?
Or some how attach it to `sizehint!`
@ararslan  may have thoughts?

Reply:
yes that's certainly possible, I did it like this allowing for advanced users to manage the memory, without impacting the basic performance. Other ways of doing it are definitely worth considering

Reply:
I don't think I like attaching it to `sizehint!` really, because I guess that implies that if I want to have index 1 million in there, that would kind of correspond to `sizehint!(1000_000)`, which in itself would mean that all pages for the first 1 million indices should be allocated wasting a lot of memory

Reply:
I am not particularly keen on attaching it to `sizehint` either.

Reply:
One possibly convoluted way is to make a IntPage datastructure that holds the page and a counter, each time something gets pushed to the page the counter gets incremented, when popped, decremented. Then the set can check whether it's 0 to clean it up. But that will lower performance somewhat I guess since extra instructions per push and pop.

Reply:
That kind of reference counting of sorts seems worth the overhead to me; having to manually manage memory is a bit unusual in Julia.

Reply:
One issue I have with something like this is that the worst-case scenario (pushing and popping the same single value a lot) is really really bad for performance. I would also really like for a way to just omit the automatic cleanup if you don't want it, maybe this can be done with two different `Page` types, and a parameter in the `SparseIntSet` which defaults to the automatically cleaned one. If this seems reasonable I think I'll have a go at that later

Reply:
On a semi side-note, is there a direct way of making an entry in a vector again `undefined` or  something like `unassign`? On what is this identification based under the hood?
That would make `cleanup!` probably a little bit better since I don't need to create a whole new page `Vector`

Reply:
I ended up sticking to a less over-engineered solution storing the counters in the set itself and adding a `dirty_pop!` that doesn't do any cleanup. There's a 30% overhead of a normal `pop!` that needs no cleaning vs a `dirty_pop!`, which is kind of a lot imo. If there are any suggestions on making this less severe that would be nice, but I don't think it can be reduced too much more.